core/drivers/clk/clk-stm32mp13.c

```diff
@@ -0,0 +1,1727 @@
+{
+	assert(osc_id >= 0 && osc_id < (int)ARRAY_SIZE(stm32mp13_osc_data));
+
+	return &stm32mp13_osc_data[osc_id];
+}
+
+static unsigned long clk_stm32_get_rate_oscillateur(int osc_id)
+{
+	struct clk_stm32_priv *priv = clk_stm32_get_priv();
+	struct stm32_clk_platdata *pdata = priv->pdata;
+	struct stm32_osci_dt_cfg *osci = &pdata->osci[osc_id];
+
+	return osci->freq;
+}
+
+static void clk_oscillator_set_bypass(struct clk_stm32_priv *priv,
+				      struct clk_oscillator_data *osc_data,
+				      bool digbyp, bool bypass)
+{
+	struct clk_stm32_bypass *bypass_data = osc_data->bypass;
+	uintptr_t address = 0;
+
+	if (!bypass_data)
+		return;
+
+	address = priv->base + bypass_data->offset;
+
+	if (digbyp)
+		io_setbits32(address, BIT(bypass_data->bit_digbyp));
+
+	if (bypass || digbyp)
+		io_setbits32(address, BIT(bypass_data->bit_byp));
+}
+
+static void clk_oscillator_set_css(struct clk_stm32_priv *priv,
+				   struct clk_oscillator_data *osc_data,
+				   bool css)
+{
+	struct clk_stm32_css *css_data = osc_data->css;
+	uintptr_t address = 0;
+
+	if (!css_data)
+		return;
+
+	address = priv->base + css_data->offset;
+
+	if (css)
+		io_setbits32(address, BIT(css_data->bit_css));
+}
+
+static void clk_oscillator_set_drive(struct clk_stm32_priv *priv,
+				     struct clk_oscillator_data *osc_data,
+				     uint8_t lsedrv)
+{
+	struct clk_stm32_drive *drive_data = osc_data->drive;
+	uintptr_t address = 0;
+	uint32_t mask = 0;
+	uint32_t value = 0;
+
+	if (!drive_data)
+		return;
+
+	address = priv->base + drive_data->offset;
+
+	mask = (BIT(drive_data->drv_width) - 1U) << drive_data->drv_shift;
+
+	/*
+	 * Warning: not recommended to switch directly from "high drive"
+	 * to "medium low drive", and vice-versa.
+	 */
+	value = (io_read32(address) & mask) >> drive_data->drv_shift;
+
+	while (value != lsedrv) {
+		if (value > lsedrv)
+			value--;
+		else
+			value++;
+
+		io_clrsetbits32(address, mask, value << drive_data->drv_shift);
+	}
+}
+
+static void stm32_enable_oscillator_hse(struct clk_stm32_priv *priv,
+					struct stm32_clk_platdata *pdata)
+{
+	struct clk_oscillator_data *osc_data = clk_oscillator_get_data(OSC_HSE);
+	struct stm32_osci_dt_cfg *osci = &pdata->osci[OSC_HSE];
+
+	if (osci->freq == 0U)
+		return;
+
+	clk_oscillator_set_bypass(priv, osc_data,  osci->digbyp, osci->bypass);
+
+	/* Enable clock and wait ready bit */
+	clk_stm32_gate_rdy_enable(osc_data->gate_id);
+
+	clk_oscillator_set_css(priv, osc_data, osci->css);
+}
+
+static void stm32_enable_oscillator_lse(struct clk_stm32_priv *priv,
+					struct stm32_clk_platdata *pdata)
+{
+	struct clk_oscillator_data *osc_data = clk_oscillator_get_data(OSC_LSE);
+	struct stm32_osci_dt_cfg *osci = &pdata->osci[OSC_LSE];
+
+	if (osci->freq == 0U)
+		return;
+
+	clk_oscillator_set_bypass(priv, osc_data, osci->digbyp, osci->bypass);
+
+	clk_oscillator_set_drive(priv, osc_data,  osci->drive);
+
+	/* Enable lse clock, but don't wait ready bit */
+	clk_stm32_enable_gate(osc_data->gate_id);
```

Comment:
should check return value? (could assert since the function always return `TEE_SUCESS`)
or maybe call straight `clk_stm32_endisable_gate(osc_data->gate_id, true);`

Reply:
i will check value